pkg/distributor/tenant_topic_tee.go

```diff
@@ -0,0 +1,395 @@
+		return fmt.Errorf("invalid strategy: %w", err)
+	}
+
+	return nil
+}
+
+// RegisterFlags adds the flags required to configure this flag set.
+func (cfg *TenantTopicConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, "distributor.tenant-topic-tee.enabled", false, "Enable the tenant topic tee, which writes logs to Kafka topics based on tenant IDs instead of using multitenant topics/partitions.")
+	f.StringVar(&cfg.TopicPrefix, "distributor.tenant-topic-tee.topic-prefix", "loki.tenant", "Prefix to prepend to tenant IDs to form the final Kafka topic name")
+	cfg.MaxBufferedBytes = 100 << 20 // 100MB
+	f.Var(&cfg.MaxBufferedBytes, "distributor.tenant-topic-tee.max-buffered-bytes", "Maximum number of bytes that can be buffered before producing to Kafka")
+	cfg.MaxRecordSizeBytes = kafka.MaxProducerRecordDataBytesLimit
+	f.Var(&cfg.MaxRecordSizeBytes, "distributor.tenant-topic-tee.max-record-size-bytes", "Maximum size of a single Kafka record in bytes")
+	f.StringVar(&cfg.Strategy, "distributor.tenant-topic-tee.strategy", "simple", "Topic strategy to use. Valid values are 'simple' or 'automatic'")
+}
+
+// PartitionResolver resolves the topic and partition for a given tenant and stream
+type PartitionResolver interface {
+	// Resolve returns the topic and partition for a given tenant and stream
+	Resolve(ctx context.Context, tenant string, totalPartitions uint32, stream KeyedStream) (topic string, partition int32, err error)
+}
+
+// SimplePartitionResolver is a basic implementation of PartitionResolver that
+// uses a prefix for topics and hash-based partitioning
+type SimplePartitionResolver struct {
+	topicPrefix string
+}
+
+// NewSimplePartitionResolver creates a new SimplePartitionResolver
+func NewSimplePartitionResolver(topicPrefix string) *SimplePartitionResolver {
+	return &SimplePartitionResolver{
+		topicPrefix: topicPrefix,
+	}
+}
+
+// Resolve implements PartitionResolver
+func (r *SimplePartitionResolver) Resolve(_ context.Context, tenant string, totalPartitions uint32, stream KeyedStream) (string, int32, error) {
+	topic := fmt.Sprintf("%s.%s", r.topicPrefix, tenant)
+	partition := int32(stream.HashKey % totalPartitions)
+	return topic, partition, nil
+}
+
+// ShardedPartitionResolver implements a partition resolution strategy that creates
+// single-partition topics for each tenant and shard combination. This approach offers
+// several advantages over traditional multi-partition topics:
+//
+// 1. Dynamic Scaling:
+//   - Topics are created in the form "<prefix>.<tenant>.<shard>"
+//   - Each topic has exactly one partition, with shards serving the same purpose
+//     as traditional partitions
+//   - Unlike traditional partitions which can only be increased but never decreased,
+//     this approach allows for both scaling up and down
+//   - When volume decreases, we can stop writing to higher-numbered shards
+//   - Old shards will be automatically cleaned up through normal retention policies
```

Comment:
Using the topic name for sharding is neat, I'm interested in seeing how it goes. 

If we ever need to, another solution would be to support "scale down" by adding a generation number to the topic, and writing to the newest generation (with some background discovery on a timer). 

That would come with its own set of problems, probably, but it would allow us to rely on Kafka's builtin partitioning more heavily. 

Either way, I don't think something like that needs to be implemented right now and I just wanted to openly share the idea. This looks good 👍